ddd/ddd-05-mem-nonblocking.DISCUSSION.md

```diff
@@ -0,0 +1,250 @@
+    so an IP packet greater than 64 KiB (using the IPv6 jumbogram header) can
+    only be sent if you have an interface with a native MTU exceeding it.
+    Another option is to use MSG_PEEK | MSG_TRUNC on systems which support it
+    but this requires calling recv twice per packet.)
+
+  - For the buffer size issue, our options for supporting existing applications
+    include:
+
+    - Using a maximum MTU of 1472 by default; then any application using
+      old BIO_read/BIO_write calls will work so long as it uses a buffer
+      of at least 1472 bytes (and there is a 1:1 correspondance of BIO calls and
+      syscalls).
+
+    - Applications which can support larger MTUs (e.g. a network with
+      jumbo frames) could indicate support via some method TBD, such as
+      capability flags (see below).
+
+    In this model, for the purposes of compatibility and minimising changes:
+
+    - BIO_ctrl_get_write_guarantee (BIO_C_GET_WRITE_GUARANTEE) would always
+      return either 0 or a value >=1472. (This is a change from BIO_dgram,
+      which does not implement BIO_C_GET_WRITE_GUARANTEE.)
+
+    - BIO_ctrl_pending (BIO_CTRL_PENDING) returns the size of the next frame
+      which will be popped by BIO_read. It's not possible to determine how many
+      frames are queued in a frame-BIO or the total amount of data queued by
+      using this function. (This is a change from BIO_dgram, which always
+      returns 0 for BIO_CTRL_PENDING.)
+
+    The following APIs could be provided for applications which want to
+    support larger MTUs (new applications, and applications willing to make
+    larger changes):
+
+    ```c
+      /*
+       * A flag settable on BIO_dgram_mem. A BIO_read call will not truncate,
+       * but instead fail if the frame is too large for the provided buffer. The
+       * application must call BIO_ctrl_pending() to determine the necessary
+       * buffer size and try again.
+       */
+      int BIO_dgram_set_no_trunc(bio, 1);
+
+      /*
+       * Returns libssl's idea of what our maximum UDP payload size is.
+       */
+      ssize_t BIO_get_max_dgram_len(BIO *bio);
+    ```
+
+  - If we do support things like variable destination addresses
+    (connection migration):
+
+    - PR #5257 is relevant prior art because it adapts BIO_dgram to support
+      variable source and destination addresses. It introduces four functions
+      taking `BIO_ADDR *`:
+
+      ```
+        BIO_get_dgram_dest   (BIO_CTRL_DGRAM_GET_ADDR)
+        BIO_get_dgram_origin (BIO_CTRL_DGRAM_GET_PEER)
+        BIO_set_dgram_dest   (BIO_CTRL_DGRAM_SET_ADDR)
+        BIO_set_dgram_origin (BIO_CTRL_DGRAM_SET_PEER)
+      ```
+
+      The get functions can be called after a BIO_read and yield information
+      on the last datagram received. Likewise, the set functions can be called
+      before a BIO_write and control the source and destination of the write.
+
+      This API should be fine.
```

Comment:
openssl/openssl#17184 introduces explicit params for the peer and local BIO_ADDR values in the proposed BIO_sendmmsg and BIO_recvmmsg functions.